plugins/package-managers/go/src/main/kotlin/utils/GoSupport.kt

```diff
@@ -25,3 +25,24 @@ package org.ossreviewtoolkit.plugins.packagemanagers.go.utils
  * https://go.dev/ref/mod#incompatible-versions.
  */
 fun normalizeModuleVersion(moduleVersion: String): String = moduleVersion.removePrefix("v").substringBefore("+")
+
+/**
+ * Module paths appear as substrings of file system paths in the module cache on the file system.
+ * Go does not rely on the file system to be case-sensitive. For this reason, Go has decided to
+ * replace every uppercase letter in file system paths with an exclamation mark followed by the
+ * letter's lowercase equivalent.
+ */
+fun escapeVersion(version: String): String {
+    require(!version.contains("!")) { "Module versions must not contain exclamation marks: $version"}
+
+    val escaped = StringBuilder()
+    for (char in version) {
+        if (char.isUpperCase()) {
+            escaped.append("!${char.lowercaseChar()}")
+        } else {
+            escaped.append(char)
+        }
+    }
+
+    return escaped.toString()
```

Comment:
Alternative implementation:

    return version.replace(Regex("[A-Z]")) {
        "!${it.value.lowercase()}"
    }

This could also extract `Regex("[A-Z]")` to a private property to the Regex is not recreated on each function call.

Reply:
Come on. Don't need a Regex for such an easy thing.

Reply:
Sure you don't. But I'd bet it's even faster (if the Regex is precompiled to a constant) and it's more readable (or at least shorter, if you're not used to Regexes) than your solution that manually iterates over each character in the string. But I'm also fine with keeping your approach.

Reply:
Okay, using a Regex now, extended the test case to make sure **all** patterns matching are replaced.